Optimize config data traversal in Store/Config/Placeholder

andrey-legayev opened the pull request:
### Description (*)
Optimizations:
 - get rid of a lot of string concatenation and then explode()
 - replace recursion with iterative algorithm to avoid data copying via parameters
 - utilize builtin array iterator functionality to avoid creation of additional variables
 - read/set data directly to current $data item, skip loops in _getData() and _setData()

My benchmarks for method process() on empty Magento cache have shown reduction in total time spent from 0.291 sec to 0.077 sec (-73%)
Tests were done on PHP 7.2 inside docker, XDebug=Off

### Fixed Issues (if relevant)
N/A

### Manual testing scenarios (*)
General regression testing

### Questions or comments
N/A

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds are green)

dmytro-ch commented on app/code/Magento/Store/Model/Config/Placeholder.php:
```diff
@@ -164,4 +206,26 @@ protected function _setValue(array &$container, $path, $value)
+     * @param string $value
+     * @return void
+     */
+    private function _setCurrentPathValue(array &$container, array $path, $value)
```
Please avoid using underscores at the beginning of the private/protected method's name since this is the legacy approach according to the PSR-2 Coding Style Guide:
```
Property names SHOULD NOT be prefixed with a single underscore to indicate protected or private visibility.
```

andrey-legayev replied:
Removed

dmytro-ch commented on app/code/Magento/Store/Model/Config/Placeholder.php:
```diff
@@ -164,4 +206,26 @@ protected function _setValue(array &$container, $path, $value)
+    /**
+     * Set array value by path
+     *
+     * @param array &$container
```
Could you please fix the static test failure on this line?
I guess the "&" is redundant here.

andrey-legayev replied:
I've copy-pasted it, now removed. Waiting for new static test results...

magento-engcom-team commented:
Hi @dmytro-ch, thank you for the review.
[ENGCOM-6133](https://jira.corp.magento.com/browse/ENGCOM-6133) has been created to process this Pull Request

andrey-legayev committed:
Optimize config data traversal in Store/Config/Placeholder

Optimizations:
 - get rid of a lot of string concatenation and then explode()
 - replace recursion with iterative algorithm to avoid data copying via parameters
 - utilize builtin array iterator functionality to avoid creation of additional variables
 - read/set data directly to current $data item, skip loops in _getData() and _setData()

My benchmarks for method process() on empty Magento cache have shown reduction in total time spent from 0.291 sec to 0.077 sec (-73%)

andrey-legayev commented:
I though a bit more on the algorithm and found a way to get rid of loops inside _getData() and _setData(). 
Data read/set directly to current $data element.
Code is complicated (a bit), but it's almost 4 times faster that original algorithm.

Added unit test for empty data.

magento-engcom-team commented:
Hi @dmytro-ch, thank you for the review.
[ENGCOM-6133](https://jira.corp.magento.com/browse/ENGCOM-6133) has been created to process this Pull Request

engcom-Alfa commented:
:heavy_check_mark: QA Passed

According to Blackfire companion:

**Before**
 
![before2](https://user-images.githubusercontent.com/51679138/67262607-1ed0d480-f4ae-11e9-8a0d-336b33fc580b.png)

**After**

![after1](https://user-images.githubusercontent.com/51679138/67262627-30b27780-f4ae-11e9-93bb-c7959e3a7cb6.png)

In my case, a reduction in total time spent from 0.443 sec to 0.215 sec (-51%)

andrey-legayev commented:
Thanks!
Results may be impacted by Blackfire. 
Initially I found this place in code by using Xdebug Profiler.
But then I measured time by adding logging directly into code without profiler to avoid it's impact:
```
$t = microtime(true);
// process()
$t = microtime(true) - $t;
echo $t;
```
